### DISCUSSION TITLE
Rope now has an LSP! Announcing the first release of pylsp-rope.

### DISCUSSION THREAD

**Discussion by lieryan** [python-rope/rope#427] [Announcements]
Hello all Rope users,

I've just finished making the first release of [pylsp-rope](https://pypi.org/project/pylsp-rope/). This is a plugin for `python-lsp-server` to extend its refactoring functionalities using Rope. I had been working on this for the past couple of weeks and it is just now starting to take shape and become usable.

If you want to try it out, you can install it with: `pip install pylsp-rope`.

This is not the first time that Rope has been implemented in a Language Server, `pylsp` itself comes with a very, very basic built-in Rope support which supports Code Completion and Rename operation powered by Rope, but this plugin will support on refactoring operations that the built-in implementation does not currently support.

Currently, `pylsp-rope` supports three refactoring operations:

- Extract variable
- Extract function/method
- Inline variable/method/parameter

But I plan to expand it to eventually support everything that Rope currently supports that can go through the LSP protocol. Hopefully this will make it much easier for people to start using Rope and refocus native Rope integrations to operations and workflows that cannot be (easily) done through LSP.

This is exciting times for Rope.

Happy Rope-ing!

**Comment by climbus** [comment]
Great news. Have your tried to run with coc vim?

**Reply by lieryan** [reply]
I was using [vim-lsp](https://github.com/prabirshrestha/vim-lsp) when writing this. I hadn't tried it with coc-vim, but I wouldn't expect there to be a massive differences between LSP clients.

Let me know of your experience if you're using other LSP clients.

**Reply by lieryan** [reply]
Ok, since I made my last comment, I've tested pylsp-rope with a couple more LSP clients, namely: ALE, Neovim native LSP client, and CoC.nvim.

Testing with ALE showed me that pylsp-rope 0.1.1 was relying on some vim-lsp specific behavior (maybe bug?) which was non-standard, so I've pushed pylsp-rope-0.1.2 to fix that. With 0.1.2, ALE now worked perfectly.

Neovim native LSP client seems to work mostly just fine. There is a couple [strange behavior](https://github.com/python-rope/pylsp-rope/issues/5), but overall it seems to do OK-ish.

Coc.nvim though doesn't seem to work at all. I've documented the issue in python-rope/pylsp-rope#4, I'll keep digging, though I'm kinda stumped with this one at the moment.

**Reply by lieryan** [reply]
Ok, I found the issue and pylsp-rope 0.1.3 should work in coc.nvim now.

**Comment by mcepl** [comment]
Is there some pluginless way how to discover/highlight/show code actions? If not, which plugin is the simplest one? I found https://www.reddit.com/r/neovim/comments/pkwl2a/underlining_spans_of_code_with_code_actions/ … does it work?

**Reply by lieryan** [reply]
If you're using Neovim 0.5 (I believe most distro is still on 4.4, I had to use an [appimage](https://github.com/neovim/neovim/releases) for that), then you should only need a plugin to configure the language server. I hadn't really researched a way to manually configure language server with Neovim native LSP yet, so I used the https://github.com/neovim/nvim-lspconfig for now to test Neovim's Native LSP configuration.

If you're using Vim or if you don't want to use the native LSP, then I find that the simplest LSP client is [prabirshrestha/vim-lsp](https://github.com/prabirshrestha/vim-lsp), weighing at just 1.2 MB. 

In comparison CoC, which I find very bloated is 208MB including its `node_modules` dependencies (+ you have to install node/yarn from system package manager). ALE is middle ground at 13M (All numbers exclude the size of the .git folder of their respective projects, so this is the number you would get when you do a shallow clone).

**Reply by mcepl** [reply]
I maintain neovim in openSUSE, so yes, [we have 0.5.1 already](https://build.opensuse.org/package/show/openSUSE:Factory/neovim) and I run [a nightly build](https://build.opensuse.org/package/show/home:mcepl:neovim/neovim) myself. So, yes, I would like to use the native neovim LSP, but I forgot that this package is not even vim-specific and even less neovim-one, so the question is not appropriate here.
